optaplanner-examples/src/main/java/org/optaplanner/examples/cheaptime/optional/score/CheapTimeConstraintProvider.java

```diff
@@ -0,0 +1,144 @@
+                .join(Resource.class,
+                        filtering((taskAssignment, resource) -> taskAssignment.getTask().getUsage(resource) > 0))
+                .join(Period.class,
+                        lessThanOrEqual((taskAssignment, resource) -> taskAssignment.getStartPeriod(), Period::getIndex),
+                        greaterThan((taskAssignment, resource) -> taskAssignment.getEndPeriod(), Period::getIndex))
+                .groupBy((taskAssignment, resource, period) -> period,
+                        (taskAssignment, resource, period) -> resource,
+                        (taskAssignment, resource, period) -> taskAssignment.getMachine(),
+                        sum((taskAssignment, resource, period) -> taskAssignment.getTask().getUsage(resource)))
+                .filter((period, resource, machine, usage) -> machine.getCapacity(resource) < usage)
+                .penalizeLong(CONSTRAINT_PACKAGE, "Maximum resource capacity", HardMediumSoftLongScore.ONE_HARD,
+                        (period, resource, machine, usage) -> usage - machine.getCapacity(resource));
+    }
+
+    protected Constraint activeMachinePowerCost(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Period.class)
+                .join(Machine.class)
+                .ifExists(TaskAssignment.class,
+                        equal((period, machine) -> machine, TaskAssignment::getMachine),
+                        greaterThanOrEqual((period, machine) -> period.getIndex(), TaskAssignment::getStartPeriod),
+                        lessThan((period, machine) -> period.getIndex(), TaskAssignment::getEndPeriod))
+                .penalizeLong(CONSTRAINT_PACKAGE, "Active machine power cost", HardMediumSoftLongScore.ONE_MEDIUM,
+                        (period, machine) -> multiplyTwoMicros(machine.getPowerConsumptionMicros(),
+                                period.getPowerPriceMicros()));
+    }
+
+    protected Constraint activeMachineSpinUpAndDownCost(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Machine.class)
+                .ifExists(TaskAssignment.class,
+                        equal(Function.identity(), TaskAssignment::getMachine))
+                .penalizeLong(CONSTRAINT_PACKAGE, "Active machine spin up and down cost", HardMediumSoftLongScore.ONE_MEDIUM,
+                        Machine::getSpinUpDownCostMicros);
+    }
+
+    protected Constraint idleCosts(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(TaskAssignment.class)
+                .groupBy(TaskAssignment::getMachine,
+                        consecutiveIntervals(TaskAssignment::getStartPeriod, TaskAssignment::getEndPeriod, (a, b) -> b - a))
+                .flattenLast(ConsecutiveIntervalInfo::getBreaks)
+                .join(Period.class,
+                        lessThanOrEqual((machine, brk) -> brk.getPreviousIntervalClusterEnd(), Period::getIndex),
```

Comment:
Stylistic note: let's turn `brk` to `break`; now we save just 2 letters and its meaning is less obvious.

Reply:
Cannot, `break` is a keyword in Java.